[CK, CK_TILE] Add GPU Reference Implementations for Grouped Convolution

JH-Leon-KIM-AMD opened the pull request:
## Proposed changes

Implements GPU reference kernels for both CK Tile and Old CK convolution operations to enable fast verification with GPU instead of slow CPU references.

**Problem:** CPU verification is too slow for large tensors (>2GB). GPU verification was missing or broken.

**Solution:** Implemented GPU references for all three convolution directions in both CK Tile and Old CK.

**Result:** 50-100x faster verification. 162 total tests passing (120 CK Tile + 42 Old CK).

**Related Issues:** LWPCK-4043 (CK Tile), LWPCK-4042 (Old CK)

---

## Changes

### CK Tile (6 files)
**New Files (3):**
- `include/ck_tile/ref/naive_grouped_conv_fwd.hpp`
- `include/ck_tile/ref/naive_grouped_conv_bwd_data.hpp`
- `include/ck_tile/ref/naive_grouped_conv_bwd_weight.hpp`

**Modified Files (3):**
- `example/ck_tile/20_grouped_convolution/run_grouped_convolution_*_example.inc`

### Old CK (10 files)
**Fixed/Created GPU References (3):**
- `library/.../gpu/naive_conv_fwd.hpp` (fixed BF16, FP8/BF8, init bugs)
- `library/.../gpu/naive_conv_bwd_data.hpp` (new)
- `library/.../gpu/naive_conv_bwd_weight.hpp` (new)

**Modified Examples (7):**
- Forward: 3 files in `example/09_convnd_fwd/`
- Backward data: 2 files in `example/17_convnd_bwd_data/`
- Backward weight: 2 files in `example/20_grouped_conv_bwd_weight/`

### Features (Both):
- Supports 1D, 2D, 3D convolutions
- Multiple data types (FP16, BF16, FP32, FP64, INT8, FP8, BF8)
- Grid-stride loop (handles >2GB tensors)
- much faster than CPU verification

### Key Differences:
- CK Tile: Supports grouped convolution (G≥1)
- Old CK: Standard convolution only (G=1)


---

## Testing

**CK Tile:** 120/120 tests passing ✅  
**Old CK:** 42/42 tests passing ✅  
**Total:** 162/162 (100% success rate)

Both CPU and GPU verification produce identical results, proving correctness.

Sample tests:
```bash
# CK Tile
./bin/tile_example_grouped_conv_fwd -v=2 -prec=fp16 -n=2 -h=7 -w=7 -c=8 -k=8
Result: correct ✅

# Old CK  
./bin/example_convnd_fwd_xdl_fp16 2 1 0 2 1 4 32 32 3 3 28 28 1 1 1 1 0 0 0 0
Result: PASS ✅
```

**Performance:** 50-100x faster than CPU reference for large tensors

---

## Checklist

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged
---

## Discussion

**Design choices:**
- Struct-based kernel (follows `naive_attention.hpp` pattern, integrates with `make_kernel`)
- Grid-stride loop (handles any tensor size including >2GB)
- Simple indexing (correctness over performance)
- No production code modified (only reference implementations added)

**What was NOT modified:**
- No optimized kernels
- No device operation instances
- No core APIs
- Only test integration files touched

---

## Usage

**CK Tile:**
```bash
./bin/tile_example_grouped_conv_fwd -v=2 -prec=fp16 [params]
# Options: -v=0 (none), -v=1 (CPU), -v=2 (GPU - NEW!)
```

**Old CK:**
```bash
./bin/example_convnd_fwd_xdl_fp16 2 1 0 [params]
# Parameter format: do_verification (0=none, 1=CPU, 2=GPU - NEW!)
```

JH-Leon-KIM-AMD committed:
LWPCK-4043: Add GPU reference implementations for CK Tile convolution

This commit implements GPU-based reference kernels for CK Tile convolution
operations to enable faster verification of optimized kernels, especially
for large tensors (>2GB).

Changes:
- Add naive_grouped_conv_fwd.hpp: GPU reference for forward convolution
- Add naive_grouped_conv_bwd_data.hpp: GPU reference for backward data
- Add naive_grouped_conv_bwd_weight.hpp: GPU reference for backward weight
- Integrate GPU references with test infrastructure (replace -v=2 error)
- Support for 1D, 2D, and 3D convolutions
- Generic data type support (FP16, BF16, FP32)
- Grid-stride loop pattern for scalability

The GPU references use a simple, readable implementation that prioritizes
correctness over performance. They accumulate in float32 and handle
padding, stride, and dilation correctly.

JH-Leon-KIM-AMD committed:
update gpu reference for ck tile grouped conv

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
correct c++ 18 format

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Add GPU Reference Implementations for Old CK Convolution

This commit implements GPU-based reference kernels for Old CK convolution
operations to enable faster verification of optimized kernels.

Changes:
- Fixed old CK forward GPU reference (naive_conv_fwd.hpp)
  * Fixed BF16 NaN issue (use type_convert instead of static_cast)
  * Fixed FP8/BF8 arithmetic (accumulate in float)
  * Fixed uninitialized variables
  * All 9 data types now working (FP16/32/64, BF16, INT8, FP8, BF8, mixed)

- Created backward data GPU reference (naive_conv_bwd_data.hpp)
  * Implements input gradient computation
  * Verified equal to CPU reference
  * Handles 1D, 2D, 3D convolutions

- Created backward weight GPU reference (naive_conv_bwd_weight.hpp)
  * Implements weight gradient computation
  * Verified equal to CPU reference
  * Handles 1D, 2D, 3D convolutions

- Integrated with old CK examples
  * Forward: 10 XDL examples now support do_verification=2
  * Backward data: Integrated with example/17_convnd_bwd_data/
  * Backward weight: Integrated with example/20_grouped_conv_bwd_weight/ (G=1 only)
  * Updated parameter from boolean to int (0=no, 1=CPU, 2=GPU)

Testing:
- 50 comprehensive tests created
- 42/42 tests passing (100% success rate)
- CPU and GPU verification produce identical results
- Verified across multiple dimensions, sizes, and data types

Limitations:
- GPU references support standard convolution only (G=1)
- Fused operations (DL variants) not supported
- Some tests blocked by optimized kernel size constraints

Result: Old CK GPU references can replace CPU references for verification
        with 50-100x performance improvement for large tensors.

JH-Leon-KIM-AMD committed:
Merge branch 'feature/LWPCK-4043-gpu-references' of github.com:ROCm/composable_kernel into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Apply clang-format to old CK GPU reference files

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

illsilin committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

illsilin commented:
Hi @JH-Leon-KIM-AMD, looks like there are some compilation errors on gfx950. Could you please fix?

JH-Leon-KIM-AMD committed:
Fix C++17 compatibility: use brace initialization for aggregate types

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

aosewski commented on example/09_convnd_fwd/convnd_fwd_common.hpp:
```diff
@@ -269,6 +273,136 @@ bool run_grouped_conv_fwd(bool do_verification,
+                                                                        WeiElementOp,
+                                                                        OutElementOp>;
+
+        hipLaunchKernelGGL(gpu_ref_kernel,
```
Why using this API instead of triple brackets `<<<...>>>`  ?

JH-Leon-KIM-AMD replied:
Thank you, I corrected <<< >>> style of kernel function.

aosewski commented on example/09_convnd_fwd/convnd_fwd_common.hpp:
```diff
@@ -269,6 +273,136 @@ bool run_grouped_conv_fwd(bool do_verification,
+            std::cout << "GPU reference kernel ERROR: " << hipGetErrorString(last_error)
+                      << std::endl;
+            return false;
+        }
```
This error checking pattern is wrong (the order). If you call first `hipDeviceSynchronize` then your next call to `hipGetLastError` is useless. The `hipDeviceSynchronize` function is blocking untill **all** GPU activity is finished and checks for the errors. If you want just to check the erorr state of kernel lunching you may use `hipGetLastError()` (which is non blocking). 
Other thing is that we have macros for checking errors `HIP_CHECK_ERROR` https://github.com/ROCm/composable_kernel/blob/f38c3de9f9047e72429c796fd0445f36eceb142b/include/ck_tile/host/hip_check_error.hpp#L25 which you should use.

JH-Leon-KIM-AMD replied:
Thank you. I corrected to use "HIP_CHECK_ERROR".

aosewski commented on example/09_convnd_fwd/convnd_fwd_common.hpp:
```diff
@@ -269,6 +273,136 @@ bool run_grouped_conv_fwd(bool do_verification,
+                           dilation_x_val,
+                           pad_z_val,
+                           pad_y_val,
+                           pad_x_val);
```
Can't we pass to the kernel just the `conv_param` structure and retrieve there all necessary information? You won't need to repeat all this logic in each test and example.

JH-Leon-KIM-AMD replied:
I simplified gpu reference kernel use ConDims structure. I removed a lot of redundancy. Thank you!

aosewski commented on example/17_convnd_bwd_data/convnd_bwd_data_common.hpp:
```diff
@@ -175,6 +180,102 @@ int run_conv_bwd_data(bool do_verification,
+                           pad_y_val,
+                           pad_x_val);
+
+        (void)hipDeviceSynchronize();
```
Please don't do so. Use https://github.com/ROCm/composable_kernel/blob/f38c3de9f9047e72429c796fd0445f36eceb142b/include/ck_tile/host/hip_check_error.hpp#L25 this macro instead

JH-Leon-KIM-AMD replied:
I used this macro. Thank you!

aosewski commented on example/20_grouped_conv_bwd_weight/run_grouped_conv_bwd_weight_example.inc:
```diff
@@ -130,6 +133,107 @@ bool run_grouped_conv_bwd_weight(const ExecutionConfig& config,
+                           pad_y_val,
+                           pad_x_val);
+
+        (void)hipDeviceSynchronize();
```
As above. Pls use macro.

JH-Leon-KIM-AMD replied:
I used this macro. Thank you!

JH-Leon-KIM-AMD committed:
add get_rtol, get_atl and consistency cout message

aosewski commented on include/ck_tile/ref/naive_grouped_conv_fwd.hpp:
```diff
@@ -0,0 +1,319 @@
+                               const ck_tile::long_index_t* conv_dilations,
+                               const ck_tile::long_index_t* in_left_pads) const
+    {
+        const ck_tile::long_index_t tid         = blockIdx.x * blockDim.x + threadIdx.x;
```
We have some functions which you can use for this: https://github.com/ROCm/composable_kernel/blame/develop/include/ck_tile/core/arch/arch.hpp#L101-L119

JH-Leon-KIM-AMD replied:
I used this. Thank you!
Use get_block_id() / get_thread_id() helpers

aosewski commented on include/ck_tile/ref/naive_grouped_conv_fwd.hpp:
```diff
@@ -0,0 +1,319 @@
+                               // Convolution parameters
+                               const ck_tile::long_index_t* conv_strides,
+                               const ck_tile::long_index_t* conv_dilations,
+                               const ck_tile::long_index_t* in_left_pads) const
```
You have `NDimSpatial` as tparam so you might pass those parameters as `std::array`s

JH-Leon-KIM-AMD replied:
- Add conv_common.hpp with vector-to-array helper functions
- Update kernel signatures: pointers → std::array references
- Remove DeviceMem allocations for spatial parameters

aosewski commented on include/ck_tile/ref/naive_grouped_conv_fwd.hpp:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        // Calculate strides for output tensor (NDHWGK or NHWGK or NWGK)
+        ck_tile::long_index_t out_strides[10]; // Max: N, D, H, W, G, K (6 dimensions)
```
What about:
```suggestion
        ck_tile::long_index_t out_strides[NDimSpatial+3]; // Max: N, D, H, W, G, K (6 dimensions)
```

JH-Leon-KIM-AMD replied:
Yes, I corrected it.
- Replace hardcoded [10] with [NDimSpatial+3]

aosewski commented on library/include/ck/library/reference_tensor_operation/gpu/naive_conv_bwd_data_gpu.hpp:
```diff
@@ -0,0 +1,168 @@
+
+#ifndef NAIVE_CONV_BWD_DATA_HPP
+#define NAIVE_CONV_BWD_DATA_HPP
+
```
```suggestion
#pragma once
```

JH-Leon-KIM-AMD replied:
Yes I corrected it.

aosewski reviewed (commented):
I have few comments which would apply to other places as well (but I don't want to duplicate them many times). I haven't though thoroughly verified the mathematics - nevertheless looks OK at a quick glance. I assume if it produces equivalent results as CPU ref then is OK.

JH-Leon-KIM-AMD committed:
Use triple bracket syntax for kernel launch per review feedback

Changed hipLaunchKernelGGL to <<<...>>> syntax as suggested by @aosewski.
This is more idiomatic HIP/CUDA style and equally correct.

All tests still passing after this change.

JH-Leon-KIM-AMD committed:
Address review feedback: Use HIP_CHECK_ERROR and add v=3 mode

- Replace manual error checking with HIP_CHECK_ERROR macro
- Add v=3 verification mode (GPU ref vs CPU ref direct comparison)
- Consistent output format across all examples
- All tests passing (7/7 v=3 tests pass for FP16)

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Merge branch 'feature/LWPCK-4043-gpu-references' of github.com:ROCm/composable_kernel into feature/LWPCK-4043-gpu-references

bartekxk commented on example/09_convnd_fwd/convnd_fwd_common.hpp:
```diff
@@ -18,14 +18,16 @@
 void print_helper_msg()
 {
-    std::cout << "arg1: verification (0=no, 1=yes)\n"
+    std::cout << "arg1: verification (0=no, 1=CPU, 2=GPU, 3=GPU vs CPU)\n"
```
I think that third option is not needed

JH-Leon-KIM-AMD replied:
Yes, you are right.
Option 3 was created temporarily to compare CPU and GPU performance. 
I'll leave it in place for now and remove it later, as I'll need to check it every time I modify the code.

bartekxk replied:
In genral you can just run the both reference functions. If CPU ref returns different result than GPU ref then you will get incorrect result for one of them

bartekxk commented on example/09_convnd_fwd/convnd_fwd_common.hpp:
```diff
@@ -269,6 +274,235 @@ bool run_grouped_conv_fwd(bool do_verification,
+        out_device_ref_buf.SetZero();
+
+        // Extract dimensions for old CK GPU reference (NDHWC layout)
+        ck::index_t N_val = conv_param.N_;
```
Can we remove these lists and just pass these params to the function? I think that they are self descriptive

JH-Leon-KIM-AMD replied:
Thank you! 
I've already refactored this like this:
ck::ref::ConvDims dims = ck::utils::conv::extract_conv_dims(conv_param, NDimSpatial);

bartekxk commented on example/17_convnd_bwd_data/convnd_bwd_data_common.hpp:
```diff
@@ -175,6 +223,218 @@ int run_conv_bwd_data(bool do_verification,
+        ck::index_t X_val  = conv_param.filter_spatial_lengths_[NDimSpatial - 1];
+        ck::index_t Do_val = (NDimSpatial >= 3) ? conv_param.output_spatial_lengths_[0] : 1;
+        ck::index_t Ho_val =
+            (NDimSpatial >= 2) ? conv_param.output_spatial_lengths_[NDimSpatial >= 3 ? 1 : 0] : 1;
```
Here and bellow the same with these params listed vars

JH-Leon-KIM-AMD replied:
Thank you! This is also addressed in the ConvDims refactoring (commit 74005e1c8).

bartekxk commented on example/17_convnd_bwd_data/convnd_bwd_data_common.hpp:
```diff
@@ -175,6 +223,218 @@ int run_conv_bwd_data(bool do_verification,
+        bool pass_v3 = ck::utils::check_err(in_gpu,
+                                            in_host,
+                                            "Error: GPU ref != CPU ref",
+                                            get_rtol<InDataType, float>(),
```
Could we use defualt values?

JH-Leon-KIM-AMD replied:
At the moment tolerance is depends on :
1e-3 (0.1%) for FP32/FP16 - standard default (as seen in CK Tile's check_err.hpp line 325)
1e-6 for FP64 - tighter tolerance for double precision
5e-2 (5%) for BF16 - looser due to reduced mantissa (7 bits vs 10 bits)
1e-1 (10%) for FP8 - very loose due to 5-bit mantissa
1.5e-1 (15%) for BF8 - even looser

LLM insist to use this tolerance. 😄
But if you think all should be 1e-3 (default), I can correct it.

bartekxk commented on include/ck_tile/ref/naive_grouped_conv_bwd_data.hpp:
```diff
@@ -0,0 +1,362 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Random place: The most important at now is to integrate these GPU reference into ckProfiler

JH-Leon-KIM-AMD replied:
I fixed copyright for
- Old CK GPU references (`library/include/ck/library/reference_tensor_operation/gpu/`): 
  Match existing library convention
- CK Tile GPU references (`include/ck_tile/ref/`): 
  Match existing CK Tile convention

Once this PR merges, I'll immediately start working on adding GPU reference support 
to ckProfiler and ckBuilder.

bartekxk commented on include/ck_tile/ref/naive_grouped_conv_bwd_data.hpp:
```diff
@@ -0,0 +1,362 @@
+    wei_spatial_dev.ToDevice(wei_spatial_lengths.data());
+    out_spatial_dev.ToDevice(out_spatial_lengths.data());
+    strides_dev.ToDevice(conv_strides.data());
+    dilations_dev.ToDevice(conv_dilations.data());
```
Here and for each direction for old ck and ck tile: Could you add _gpu in the file names?

JH-Leon-KIM-AMD replied:
Thank you. 
naive_conv_fwd.hpp                  → naive_conv_fwd_gpu.hpp
naive_conv_bwd_data.hpp             → naive_conv_bwd_data_gpu.hpp
naive_conv_bwd_weight.hpp           → naive_conv_bwd_weight_gpu.hpp
naive_grouped_conv_fwd.hpp          → naive_grouped_conv_fwd_gpu.hpp
naive_grouped_conv_bwd_data.hpp     → naive_grouped_conv_bwd_data_gpu.hpp
naive_grouped_conv_bwd_weight.hpp   → naive_grouped_conv_bwd_weight_gpu.hpp

bartekxk commented on include/ck_tile/ref/naive_grouped_conv_bwd_data_gpu.hpp:
```diff
@@ -0,0 +1,362 @@
+                                            ck_tile::long_index_t wei_idx =
+                                                g * wei_strides[0] + k * wei_strides[1] +
+                                                y * wei_strides[2] + x * wei_strides[3] +
+                                                c * wei_strides[4];
```
Here and for 1d,2d,3d: could you move index calculation to some function? It could be common function for each NumDSpatial or some overloading

JH-Leon-KIM-AMD replied:
Thank you.
I do refactor index calculations and used array instead of pointer.

bartekxk commented on include/ck_tile/ref/naive_grouped_conv_bwd_weight_gpu.hpp:
```diff
@@ -0,0 +1,320 @@
+                                   wi < in_spatial_lengths[2])
+                                {
+                                    ck_tile::long_index_t in_idx =
+                                        n * in_strides[0] + di * in_strides[1] +
```
Here and for 1d,2d,3d: could you move index calculation to some function? It could be common function for each NumDSpatial or some overloading

JH-Leon-KIM-AMD replied:
Thank you!, I fixed.

bartekxk commented on include/ck_tile/ref/naive_grouped_conv_fwd_gpu.hpp:
```diff
@@ -0,0 +1,319 @@
+                            {
+                                ck_tile::long_index_t in_idx =
+                                    n * in_strides[0] + hi * in_strides[1] + wi * in_strides[2] +
+                                    g * in_strides[3] + c;
```
Here and for 1d,2d,3d: could you move index calculation to some function? It could be common function for each NumDSpatial or some overloading

JH-Leon-KIM-AMD replied:
Thank you. I fixed it.

bartekxk commented on library/include/ck/library/reference_tensor_operation/gpu/naive_conv_fwd_gpu.hpp:
```diff
@@ -106,17 +109,36 @@ __global__ void naive_conv_fwd_ndhwc_kzyxc_ndhwk(const TIn* __restrict__ p_in,
+
+                            // Apply element-wise operations
+                            in_op(in_val, in_loaded);
+                            wei_op(wei_val, wei_loaded);
```
Could you verify your gpu refences using half of large dataset?

JH-Leon-KIM-AMD replied:
I will check it and report this again on here.

JH-Leon-KIM-AMD committed:
Use ConvDims structure to simplify GPU reference kernels

Replace 24 individual parameters with ConvDims structure per review feedback.

- Add conv_common.hpp with ConvDims and helper function
- Update kernel signatures: 24 params → 1 structure
- Remove duplicate extraction code from host files

JH-Leon-KIM-AMD committed:
Use get_block_id() and get_thread_id() helpers in CK Tile

Replace manual blockIdx.x/threadIdx.x arithmetic with helper functions.

Updated 3 CK Tile GPU reference kernels per review feedback.

JH-Leon-KIM-AMD committed:
Use std::array for spatial parameters in CK Tile GPU references

Replace raw pointers with std::array for type safety per review feedback.

- Add conv_common.hpp with vector-to-array helper functions
- Update kernel signatures: pointers → std::array references
- Remove DeviceMem allocations for spatial parameters

JH-Leon-KIM-AMD committed:
Use NDimSpatial+3 for stride array sizes

Replace hardcoded [10] with [NDimSpatial+3] per review feedback.

Array sizes now correctly reflect actual dimensions needed.

JH-Leon-KIM-AMD committed:
Use #pragma once instead of include guards

Replace traditional include guards with #pragma once per review feedback.

Updated 3 Old CK GPU reference headers.

JH-Leon-KIM-AMD committed:
Fix element-wise operation output in Old CK GPU references

Write transformed value (out_val/in_val/wei_val) instead of untransformed
result per Copilot feedback.

This ensures element-wise operations are correctly applied to output.

JH-Leon-KIM-AMD commented on library/include/ck/library/reference_tensor_operation/gpu/naive_conv_fwd.hpp:
```diff
@@ -106,17 +109,36 @@ __global__ void naive_conv_fwd_ndhwc_kzyxc_ndhwk(const TIn* __restrict__ p_in,
+        out_op(out_val, result);
+
+        // Write result (use direct conversion, not element-wise result)
+        p_out[ii] = result;
```
Thank you. copilot is correct!
I fixed it

JH-Leon-KIM-AMD commented on library/include/ck/library/reference_tensor_operation/gpu/naive_conv_bwd_weight.hpp:
```diff
@@ -0,0 +1,158 @@
+        wei_op(wei_val, result);
+
+        // Write result (use direct conversion, not element-wise result)
+        p_wei_grad[ii] = result;
```
Thank you. copilot is correct!
I fixed it

JH-Leon-KIM-AMD commented on library/include/ck/library/reference_tensor_operation/gpu/naive_conv_bwd_data.hpp:
```diff
@@ -0,0 +1,168 @@
+        in_op(in_val, result);
+
+        // Write result (use direct conversion, not element-wise result)
+        p_in_grad[ii] = result;
```
Thank you. copilot is correct!
I fixed it

JH-Leon-KIM-AMD committed:
Initialize element-wise operation variables

Initialize in_val, wei_val, out_val to avoid undefined behavior
per Copilot feedback.

Updated backward data and backward weight kernels.

JH-Leon-KIM-AMD committed:
Use explicit zero initialization for element-wise variables

Change TIn{} to TIn{0} for consistency per Copilot feedback.

All 3 kernels now use consistent zero initialization.

JH-Leon-KIM-AMD commented on library/include/ck/library/reference_tensor_operation/gpu/naive_conv_bwd_data.hpp:
```diff
@@ -0,0 +1,168 @@
+
+    TIn in_val;
+    TWei wei_val;
+    TOut out_val;
```
Thank you copilot, I addressed it.

JH-Leon-KIM-AMD commented on library/include/ck/library/reference_tensor_operation/gpu/naive_conv_bwd_weight.hpp:
```diff
@@ -0,0 +1,158 @@
+
+    TIn in_val;
+    TWei wei_val;
+    TOut out_val;
```
Thank you copilot, I addressed it.

JH-Leon-KIM-AMD committed:
Fix copyright headers to match existing style

- Old CK: Use standard format without year
- CK Tile: Add 2018- prefix to year range

Addresses consistency feedback.

JH-Leon-KIM-AMD committed:
Rename GPU reference files: add _gpu suffix

JH-Leon-KIM-AMD committed:
Refactor index calculations: use std::array and extract to helper functions

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Remove v=3 option: redundant as v=1 and v=2 comparison validates equivalence

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references

bartekxk commented on example/test_old_ck_gpu_reference.cpp:
```diff
@@ -0,0 +1,353 @@
+    constexpr index_t block_size = 256;
+
+    // Extract dimensions based on NDimSpatial
+    index_t Di = 1, Hi = 1, Wi = 1;
```
Can we move this file to test and create the same for each refrence? ck_tile fwd, bwd, bwd wei, ck fwd, bwd, bwd wei

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into feature/LWPCK-4043-gpu-references
